[트랜디 심원두] Modify: ProductListView filtering, ordering 수정. SearchView 병합.

pm1100tm opened the pull request:
불필요한 파일 삭제 후 다시 pull request 하였습니다.

kim-jeongsik committed:
[트렌디 김정식]
1) Refactor: 모델링 피드백 models.py 최종 반영
2) Add: 상품리뷰 CRUD 구현을 위한 Review views.py 및 urls.py 작성(프론트와
    API test 완료)

kim-jeongsik committed:
Fixed untracked files

kim-jeongsik committed:
Fixed untracked files

kim-jeongsik committed:
Refactor: models.py, views.py 코멘트 반영 완료

pm1100tm committed:
[트랜디 심원두] Add: ProductListView
views.py
urls.py
카테고리별, 서브 카테고리별, 판매량 순 상품 리스트 기능 작성

pm1100tm committed:
[트랜디 심원두] delete unnessesary folder

pm1100tm committed:
[트랜디 심원두] Add: ProductDetailView 작성

kim-jeongsik committed:
Add: Order OrderListView 작성, Update: Order OrderList 모델 추가

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,146 @@
+    def get(self, request, product_id):
+        
+        try:
+            if not type(product_id) is int:
```
이 역할을 하는 파이썬 함수가 있습니다! :) 한 번 써보세요~

pm1100tm replied:
소헌님, 리뷰 해주셔서 감사합니다 ! 
- isinstance(object, classinfo) 를 사용하도록 바꾸었습니다~!

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,146 @@
+            
+            product = Product.objects.get(id=product_id)
+            
+            product_detail_images = product.productdetailimage_set.filter(product=product)
```
이미 `product` 객체를 83번 줄에서 지칭 하셨기 때문에 이 객체의 productdetailimage_set은 모두 해당 product의 이미지들입니다. 따라서 product로 필터링을 다시 해줄 필요가 없습니다 :)

pm1100tm replied:
그렇군요..! 감사합니다.
아래 코드로 바꾸었습니다.

product_detail_images = product.productdetailimage_set.all()

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,146 @@
+            
+            detail_images = []
+            if product_detail_images:
+                detail_images = [images.detail_image_url for images in product_detail_images]
```
리스트 컴프리헨션은 선언 하는 그 자체만으로도 리스트를 생성해주기 때문에 87번 줄에서 리스트를 미리 만들어줄 필요가 없습니다.

pm1100tm replied:
아하..! 감사합니다. 아래와 같이 수정하였습니다.
detail_images = [
                images.detail_image_url for images in product_detail_images
            ]

테스트 해보았고, 이미지가 없을 때 빈 리스트가 출력되었습니다. 역시..

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,146 @@
+            if product_detail_images:
+                detail_images = [images.detail_image_url for images in product_detail_images]
+            
+            reviews = product.review_set.filter(product=product)
```
마찬가지로 이미 product 객체를 83번 줄에서 지칭 하셨기 때문에 이 객체의 review_set은 모두 해당 product의 리뷰들입니다. 따라서 product로 필터링을 다시 해줄 필요가 없습니다 :)

pm1100tm replied:
수정하였습니다 !
reviews = product.review_set.all()

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,146 @@
+            
+            reviews = product.review_set.filter(product=product)
+            
+            review_points_avg = 0
+            if reviews:
+                total_review_point = 0
+                for review in reviews:
+                    total_review_point += review.star
```
특정해온 리뷰들의 별점(review.star)을 모두 합치는 과정 맞나요?
이를 위해 장고에서 제공하는 메소드가 있습니다. 이 상황을 한 번 검색해보세요 :)
추천 검색어는... `how to sum certain field in django model?` 정도..?

pm1100tm replied:
바로 그 과정 맞습니다..! 
장고에서 제공하는 Sum 이라는 메서드를 사용하여 보았습니다.
avg_reviews_point = reviews.aggregate(Sum('star')).get('star__sum') / reviews.count()

그리고 `how to sum certain field in django model?` 로 검색을 해보았는데, 리스트 컴프리헨션이 더 빠르다는 이야기가 있어서 다시 바꾸어봤습니다.

avg_reviews_point = round(sum([review.star for review in reviews]) / reviews.count())

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,146 @@
+                
+                review_points_avg = get_avg_review_point(total_review_point, reviews.count())
+            
+            colors      = product.productcolor_set.filter(product=product)
+            color_names = [color.color.name for color in colors]
```
이 두 과정을 따로 진행하시지 않아도 됩니다 :)

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,146 @@
+            
+            colors      = product.productcolor_set.filter(product=product)
+            color_names = [color.color.name for color in colors]
+            sizes       = product.productsize_set.filter(product=product)
+            size_name   = [size.size.name for size in sizes]
```
이 두 과정 또한 따로 진행하시지 않아도 됩니다 :)

pm1100tm replied:
수정하였습니다..!

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,146 @@
+            sizes       = product.productsize_set.filter(product=product)
+            size_name   = [size.size.name for size in sizes]
+            
+            if product:
```
`product`는 `get( )` 메소드로 선언하셨기 때문에, 데이터가 없으면 바로 `DoesNotExist` 에러가 발생합니다. 반대로 데이터가 있다면 바로 객체가 가져와지죠. 따라서 `if product`는 언제나 참이 될것입니다. 그래서 여기서 가정해주지 않으셔도 됩니다 :)

pm1100tm replied:
!!

꼼꼼하게 봐주셔서 감사합니다~! 수정하였습니다

soheon-lee commented on review/views.py:
```diff
@@ -1,3 +1,110 @@
+class ReviewView(View):
+    def post(self, request):
+        data=json.loads(request.body)
+        print("------request.body", request.body)
```
사용하지 않는 print 문은 제거해주세요

pm1100tm replied:
앗.. 리뷰는 정식님의 코드인데 올라갔나본데요..

pm1100tm replied:
앗 이건 정식님 담당 부분인데, branch 끼리 머지 하면서 올라간 것 같습니다.

soheon-lee commented on review/views.py:
```diff
@@ -1,3 +1,110 @@
+        data=json.loads(request.body)
+        print("------request.body", request.body)
+        try:
+            user             = User.objects.get(pk=data['user_id'])
+            product          = Product.objects.get(pk=data["product_id"])
```
리뷰를 생성할 때에는 `user_id`, `product_id`를 바로 사용하실 수 있기 때문에, Review 객체를 생성할 때 id를 바로 사용하신다면 User 와 Product 객체를 굳이 가져오지 않으셔도 되기 때문에 query를 두 개 아끼실 수 있습니다 ! :) 쿼리 절약

soheon-lee commented on review/views.py:
```diff
@@ -1,3 +1,110 @@
+
+            if not reviews.exists():
+                return JsonResponse({"message": "There are not any reviews"}, status=200)
+            review_list = []
+            for review in reviews:
+                review_data = {
+                    "content"   : review.content,
+                    "image_url" : review.image_url,
+                    "star"      : review.star,
+                    "nick_name" : review.user.nick_name,
+                    "product"   : review.product_id,
+                    "updated_at": str(review.updated_at),
+                    # "user_information": review.user_information
+                }
+                review_list.append(review_data)
```
여기로 list comprehension으로 한 번 ㅂㅏ꿔보세요 :)

pm1100tm committed:
[트랜디 심원두] Update: 리뷰 반영.
product/views.py: ProductDetailView, ProductListView()

kim-jeongsik committed:
Fix: Merge conflict 문제 해결

kim-jeongsik committed:
Fix: Merge conflict 문제 해결_2

kim-jeongsik committed:
Fix: Merge conflict 문제 해결_2

pm1100tm committed:
[트렌디 심원두] intermidiated_commit before merge for insert data.

pm1100tm committed:
[트렌디 심원두] refactoring: ProductListView - intermediated commit.

pm1100tm committed:
intermeditate commit: modified ProductListView

pm1100tm committed:
intermediate commit for merge ProductListView with SearchView.

pm1100tm committed:
[트랜디 심원두] Modified ProductListView for pull request.
1. ProductListView 안에 재훈의 SearchView 코드 merge.
2. ProductListView filtering, ordering 기능 수정.

lordmyshepherd commented on product/views.py:
```diff
@@ -1,29 +1,199 @@
+                    'orderlist_set',
+                    'review_set'
+                ).\
+                annotate(sum=Sum('orderlist__quantity')).\
+                annotate(review_count=Subquery(
+                    Review.objects.filter(product=OuterRef('pk')).values('product').
+                        annotate(count=Count('pk')).values('count'))
+                )
```
👍

pm1100tm replied:
이거 왜 돌아가는지 모르겠어요 조사해보고 스택오버 플로우 업데이트 할께요, 리뷰해주셔서 감사합니다 소헌 멘토님 !

soheon-lee commented on product/views.py:
```diff
@@ -1,29 +1,199 @@
+            if not isinstance(product_id, int):
+                raise TypeError
+            
+            product = Product.objects.get(id=product_id)
+            
+            product_detail_images = product.productdetailimage_set.all()
+            
+            detail_images = [
+                images.detail_image_url for images in product_detail_images
+            ]
+            
+            reviews = product.review_set.all()
+            avg_review_point = round(
+                sum([review.star for review in reviews]) / reviews.count()
+            )
+            
+            colors  = [color.color.name
+                       for color in product.productcolor_set.all()]
+            sizes   = [size.size.name
+                       for size in product.productsize_set.all()]
+            
```
원두님 대박 여기서는 140번 프로덕트 객체에서부터 모든 데이터를 가져오셨네요 정말 굿굿입니다 !!

pm1100tm replied:
후후

lordmyshepherd commented:
LGTM 👍
